### PR TITLE
Configurable Authorizer Type

### LINKED ISSUE #4325: Change Authorizer type to "REQUEST" (instead of "TOKEN")
# This is a Feature Proposal
## Description
Change Authorizer type to "REQUEST" (instead of "TOKEN") so that the Authorizer function receives all of the parameters passed to the main function (through lamba-proxy).

As-is, it is not possible to use any parameters in the Authorizer function (other than a single authorization header) which greatly restricts what is possible in the Authorizer.

See https://forum.serverless.com/t/set-custom-authorizer-type-to-request-not-token/2760/3

## Additional Data

* ***Serverless Framework Version you're using***:
* ***Operating System***:
* ***Stack Trace***:
* ***Provider Error messages***:

### PR DESCRIPTION
**Pull request by jspies**
## What did you implement:

Closes #4325

I noticed this PR was sitting https://github.com/serverless/serverless/pull/4339. I contacted @realmike33 to see if I could help.

## How did you implement it:

Check for a passed in type value and assign. If empty, default to 'TOKEN'.

## How can we verify it:

Check that an Authorizer was created with type Request instead of Token.

screen shot 2017-10-10 at 1 47 59 pm

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -154,104 +154,108 @@ This table is generated from https://github.com/serverless/plugins/blob/master/p
 -->
 | Plugin | Author |
 |:-------|:------:|
-| **[Raml Serverless](https://github.com/andrewcurioso/raml-serverless)** <br/> Serverless plugin to work with RAML API spec documents | [andrewcurioso](http://github.com/andrewcurioso) |
```
Comment (RafalWilinski) [inline README.md]: Any idea why is this updated? 🤔

File: docs/providers/aws/events/apigateway.md [hunk 2]
```diff
@@ -313,6 +314,24 @@ functions:
             identityValidationExpression: someRegex
 ```
 
+You can also use the Request Type Authorizer by setting the `type` property. In this case, your `identitySource` could contain multiple entries for you policy cache.
```
Comment (RafalWilinski) [inline docs/providers/aws/events/apigateway.md]: Maybe we could add a note saying that `token` is default type? It's not obvious without digging into the code.

File: README.md (hunk 1 above)
Comment (jspies) [reply README.md]: I ran `npm run docs`. I figured updating the docs required this. I can roll back if needed.

File: docs/providers/aws/events/apigateway.md (hunk 2 above)
Comment (jspies) [reply docs/providers/aws/events/apigateway.md]: sure

**Comment by RafalWilinski** [comment]
Thanks for really quick update @jspies! It's LGTMized now even more 😆

**Comment by ozbillwang** [comment]
@jspies 

Thanks to raising and get this PR merged.  It is just the right time to have this feature.

In request type, there are four choices:

```
NONE
Validate boby
Validate body, query string parameters, and headers
Validate query string parameters and headers
```

screen shot 2017-11-10 at 7 31 18 am

With current request type feature, I can’t find out where to set with above options. Then I need to know what option currently it supported.

Any hint for me?

**Comment by jspies** [comment]
@ozbillwang This is a different field. You are looking at Request Validator. This PR addresses the Request Type on an Authorizer lambda, which affects which Request data is passed into the Authorizer.

I don't have much experience with the Request Validator and I've never used one in the Serverless Framework. Sorry I'm not much help.

**Pull request merged**